IBX-4647: Added WebpFormatVariationPathGenerator handling .webp images

webhdx opened the pull request:
| Question                                  | Answer
| ---------------------------------------- | ------------------
| **JIRA issue**                          | [IBX-4647](https://issues.ibexa.co/browse/IBX-4647)
| **Type**                                   | bug
| **Target Ibexa version** | `v3.3`
| **BC breaks**                          | no

`liip/imagine` package has support for generating `.webp` images. You can explicitly define `format: webp` for image variation and it'll encode the image + serve it with `.webp` extension. This isn't supported on our side. Even if you use `format: webp` configuration, the image is encoded properly but it's still returned with original file extension. This PR adds new `VariationPathGenerator` that checks for webp format and applies `.webp` file extension.

#### Checklist:
- [x] Provided PR description.
- [x] Tested the solution manually.
- [x] Provided automated test coverage.
- [x] Checked that target branch is set correctly (master for features, the oldest supported for bugs).
- [x] Ran PHP CS Fixer for new PHP code (use `$ composer fix-cs`).
- [x] Asked for a review (ping `@ezsystems/engineering-team`).

webhdx committed:
IBX-4647: Added WebpFormatVariationPathGenerator handling .webp images

alongosz commented on eZ/Bundle/EzPublishCoreBundle/Imagine/VariationPathGenerator/WebpFormatVariationPathGenerator.php:
```diff
@@ -0,0 +1,42 @@
+ * @copyright Copyright (C) Ibexa AS. All rights reserved.
+ * @license For full copyright and license information view LICENSE file distributed with this source code.
+ */
+namespace eZ\Bundle\EzPublishCoreBundle\Imagine\VariationPathGenerator;
```
Please move it to the new namespace. Otherwise you'd have to add current namespace to the compatibility layer.

alongosz commented on eZ/Bundle/EzPublishCoreBundle/Imagine/VariationPathGenerator/WebpFormatVariationPathGenerator.php:
```diff
@@ -0,0 +1,42 @@
+/**
+ * @copyright Copyright (C) Ibexa AS. All rights reserved.
+ * @license For full copyright and license information view LICENSE file distributed with this source code.
+ */
```
Missing strict_types declaration

alongosz commented on eZ/Bundle/EzPublishCoreBundle/Imagine/VariationPathGenerator/WebpFormatVariationPathGenerator.php:
```diff
@@ -0,0 +1,42 @@
+            return $variationPath;
+        }
+
+        return $variationPath . '.webp';
```
won't this result with e.g. `image.jpg.webp` instead of `image.webp`

webhdx replied:
This is correct. It looks weird but `webp` format actually uses double extension: https://symfony.com/bundles/LiipImagineBundle/current/basic-usage.html#use-webp-if-supported

alongosz commented on eZ/Bundle/EzPublishCoreBundle/Resources/config/image.yml:
```diff
@@ -285,5 +285,11 @@ services:
+        decorates: ezpublish.image_alias.variation_path_generator
+        arguments:
+            $innerVariationPathGenerator: '@.inner'
+            $filterConfiguration: "@liip_imagine.filter.configuration"
```
CS: quoting inconsistency
```suggestion
            $filterConfiguration: '@liip_imagine.filter.configuration'
```

alongosz commented on eZ/Bundle/EzPublishCoreBundle/Tests/Imagine/VariationPathGenerator/WebpFormatVariationPathGeneratorTest.php:
```diff
@@ -0,0 +1,75 @@
+ * @copyright Copyright (C) Ibexa AS. All rights reserved.
+ * @license For full copyright and license information view LICENSE file distributed with this source code.
+ */
+namespace eZ\Bundle\EzPublishCoreBundle\Tests\Imagine\VariationPathGenerator;
```
the same for test classes - they need to be in `tests/` in a corresponding namespace.

alongosz commented on eZ/Bundle/EzPublishCoreBundle/Tests/Imagine/VariationPathGenerator/WebpFormatVariationPathGeneratorTest.php:
```diff
@@ -0,0 +1,75 @@
+        );
+
+        self::assertEquals(
+            'tmp/variation/test.jpeg.webp',
```
Is it really correct? Not `test.webp`?

webhdx replied:
Yes. WebP format can work with JPEGs, GIFs and PNGs.

alongosz commented on eZ/Bundle/EzPublishCoreBundle/Tests/Imagine/VariationPathGenerator/WebpFormatVariationPathGeneratorTest.php:
```diff
@@ -0,0 +1,75 @@
+        $this->filterConfiguration = $this->createMock(FilterConfiguration::class);
+    }
+
+    public function testGetVariationPath()
```
CS
```suggestion
    public function testGetVariationPath(): void
```

alongosz commented on eZ/Bundle/EzPublishCoreBundle/Tests/Imagine/VariationPathGenerator/WebpFormatVariationPathGeneratorTest.php:
```diff
@@ -0,0 +1,75 @@
+        );
+    }
+
+    public function testGetVariationNonWebpVariation()
```
CS:
```suggestion
    public function testGetVariationNonWebpVariation(): void
```

webhdx committed:
IBX-4647: Addressed review comments

webhdx commented:
Fixed all remarks: [7e76a66](https://github.com/ezsystems/ezplatform-kernel/pull/361/commits/7e76a6682e00a428aeb897fec227ac97c139260a) I moved `WebpFormatVariationPathGenerator` to `Ibexa\Core\Imagine\VariationPathGenerator` namespace.

webhdx committed:
IBX-4647: Moved WebpFormatVariationPathGenerator to src/bundle/Core

konradoboza commented on src/bundle/Core/Imagine/VariationPathGenerator/WebpFormatVariationPathGenerator.php:
```diff
@@ -0,0 +1,44 @@
+        $this->filterConfiguration = $filterConfiguration;
+    }
+
+    public function getVariationPath($originalPath, $filter)
```
Won't specifying parameters' and return types still work despite lack of this typing in the interface itself?

alongosz replied:
> Won't specifying parameters' and return types still work despite lack of this typing in the interface itself?

If a return type is well known or well predictable, then imo we should specify it here.

webhdx committed:
IBX-4647: Added type hints

webhdx committed:
fixup! IBX-4647: Added type hints

Steveb-p commented on composer.json:
```diff
@@ -114,7 +114,8 @@
         "allow-plugins": {
-            "composer/package-versions-deprecated": true
+            "composer/package-versions-deprecated": true,
+            "php-http/discovery": false
```
```suggestion
            "*": false
```

alongosz replied:
Please wait for #362 to be merged.

Steveb-p commented on tests/bundle/Core/Imagine/VariationPathGenerator/WebpFormatVariationPathGeneratorTest.php:
```diff
@@ -0,0 +1,77 @@
+
+namespace Ibexa\Tests\Bundle\Core\Imagine\VariationPathGenerator;
+
+use eZ\Bundle\EzPublishCoreBundle\Imagine\Filter\FilterConfiguration;
+use eZ\Bundle\EzPublishCoreBundle\Imagine\VariationPathGenerator;
+use Ibexa\Bundle\Core\Imagine\VariationPathGenerator\WebpFormatVariationPathGenerator;
+use PHPUnit\Framework\TestCase;
+
+final class WebpFormatVariationPathGeneratorTest extends TestCase
+{
+    /** @var \eZ\Bundle\EzPublishCoreBundle\Imagine\VariationPathGenerator|\PHPUnit\Framework\MockObject\MockObject */
+    private $innerVariationPathGenerator;
+
+    /** @var \eZ\Bundle\EzPublishCoreBundle\Imagine\Filter\FilterConfiguration|\PHPUnit\Framework\MockObject\MockObject */
+    private $filterConfiguration;
```
Some old namespaces are used here.

Steveb-p commented on src/bundle/Core/Imagine/VariationPathGenerator/WebpFormatVariationPathGenerator.php:
```diff
@@ -0,0 +1,44 @@
+
+namespace Ibexa\Bundle\Core\Imagine\VariationPathGenerator;
+
+use eZ\Bundle\EzPublishCoreBundle\Imagine\VariationPathGenerator;
+use Liip\ImagineBundle\Imagine\Filter\FilterConfiguration;
+
+/**
+ * Decorates VariationPathGenerator with .webp extension if image variation is configured for this format.
+ */
+final class WebpFormatVariationPathGenerator implements VariationPathGenerator
+{
+    /** @var \eZ\Bundle\EzPublishCoreBundle\Imagine\VariationPathGenerator */
+    private $innerVariationPathGenerator;
```
Old namespaces.

alongosz replied:
@Steveb-p for existing classes we don't have forward compat-layer in 3.3, it needs to be fixed on a merge up. We place new classes under the new namespaces to avoid updating compat-layer.

adamwojs commented:
Could you please merge up changes @webhdx ?
